# Kotlin 使用资源进行尝试

> 原文:[https://www.educba.com/kotlin-try-with-resources/](https://www.educba.com/kotlin-try-with-resources/)

![Kotlin try-with-resources](../Images/3a28db104811bc23485d8ea4f445548b.png)

<noscript><img class="alignnone size-full wp-image-546047" src="../Images/3a28db104811bc23485d8ea4f445548b.png" alt="Kotlin try-with-resources" width="900" height="500" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/Kotlin-try-with-resources.jpg 900w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/Kotlin-try-with-resources-300x167.jpg 300w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/Kotlin-try-with-resources-768x427.jpg 768w" sizes="(max-width: 900px) 100vw, 900px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/Kotlin-try-with-resources.jpg"/></noscript>

## Kotlin 带资源尝试简介

下面的文章提供了 Kotlin try-with-resources 的概要。Kotlin try with resources 用于管理 Kotlin 中的内存，kotlin 以非常高效的方式管理内存。按照 kotlin 中 java 开发人员的说法，我们不需要显式地管理内存。在 kotlin 中，我们使用资源尝试技术来管理内存。在 kotlin 语言中，我们将函数作为 use，它负责自动管理资源。

kotlin 使用资源的尝试是 kotlin 提供的标准库功能的一部分。根据 kotlin 的文档，use 函数被定义为所有类型的可封闭的泛型的扩展。在使用函数中，定义块是处理可关闭资源的函数。基本上在 kotlin 中，try with resources 是一个 try 语句，用于声明一个或多个资源。资源只不过是一个对象，在我们的程序中，一旦使用它，它就会被关闭。例如，套接字连接资源。

<small>网页开发、编程语言、软件测试&其他</small>

### 关键要点

*   kotlin 中的 Try with resource 语句将允许我们提供一组在代码完成执行后自动关闭的资源。
*   我们不需要手动关闭文件。我们在 kotlin 中使用 use 函数，而不是 java 中使用的这个函数。

### 为什么使用 Kotlin try-with-resources？

try with resource 语句将确保每个资源在语句执行结束时关闭。假设我们没有关闭资源，这将构成资源泄漏，我们的程序也在耗尽它可用的资源。我们将一个对象作为资源传递，它实现了可自动关闭的对象，包括通过使用一个可关闭的对象实现的对象。通过使用 kotlin try with resource，我们不需要添加额外的 finally 块，只需传递资源的结束语句。

在出现异常块时，try、catch 和 finally 块与带资源的 try 块是不同的。假设从 try 和 finally 块都抛出了异常，那么从 try 和资源一起抛出的异常就是被抑制的异常。

下面的例子说明了我们为什么要对资源使用 try，如下所示:

**代码:**

```
fun main (args: Array<String>) {
val trywith = File ("kotlin.txt")
trywith.bufferedReader ().use {
println (it.readText())
}
}
```

**输出:**

![kotlin try-with-resources 1](../Images/ed3af0c68851c8ea501a355f2032e605.png)

<noscript><img class="alignnone wp-image-545675 size-full" src="../Images/ed3af0c68851c8ea501a355f2032e605.png" alt="kotlin try-with-resources 1" width="694" height="259" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-1.jpg 694w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-1-300x112.jpg 300w" sizes="(max-width: 694px) 100vw, 694px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-1.jpg"/></noscript>

在上面的例子中，我们使用了 main 函数，在 main 函数中，我们定义了数组字符串，我们还使用了 bufferedReader，并使用了带有变量 trywith 的函数。

### 科特林尝试资源管理

在使用 kotlin try with resources 时，我们可以定义三个不同的阶段。以下示例显示了 kotlin 对资源管理的尝试，如下所示:

**代码:**

```
fun main(args: Array<String>)
{
  res = acquireResource()
try
{
  useResource(res)
} finally {
  releaseResource(res)
 }
}
```

**输出:**

![management](../Images/5f277481689587843a32fedaf3561b29.png)

<noscript><img class="alignnone wp-image-545676 size-full" src="../Images/5f277481689587843a32fedaf3561b29.png" alt="management" width="728" height="269" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-2.jpg 728w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-2-300x111.jpg 300w" sizes="(max-width: 728px) 100vw, 728px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-2.jpg"/></noscript>

如果是语言库负责资源的释放，那么我们称之为自动资源管理。这种类型的特征将我们从记忆免费资源中解放出来。资源管理通常会受到范围块的限制。假设我们同时处理不止一个资源，它们会以一种良好的方式被释放。

在持有资源的 java 对象中，它有资格实现对特定接口自动进行的资源管理。在 kotlin 中，有相同的资源持有者的概念，对象将从这些资源持有者实现自动可关闭或可关闭。

### 如何使用 kotlin 资源功能？

为了自动管理资源，许多语言都有自己的专用结构，有时它们会提供一个模式或库方法。根据设计，我们可以在标准库中找到用于调用使用函数的扩展方法。

下面的例子显示了我们如何使用 kotlin 资源中的函数，如下所示:

**代码:**

```
fun main(args: Array<String>)
{
  val kot_fun = FileWriter ("kotlin.txt")
kot_fun.use
{
  kot_fun.write ("kotlin")
}
}
```

**输出:**

![how we can use function](../Images/6282bbc3ac2bb1a3cf6cb3ff8ed4bb20.png)

<noscript><img class="alignnone wp-image-545677 size-full" src="../Images/6282bbc3ac2bb1a3cf6cb3ff8ed4bb20.png" alt="how we can use function" width="800" height="260" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-4.jpg 800w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-4-300x98.jpg 300w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-4-768x250.jpg 768w" sizes="(max-width: 800px) 100vw, 800px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-4.jpg"/></noscript>

我们通过对资源使用 try 来调用任何实现 closeable 和 auto closeable 方法的对象上的 use 函数。use 方法将获取 lambda 表达式，执行它，每当执行将通过使用异常或通过正常使用离开块时，它将释放资源。

在下面的例子中，我们通过创建闭包来使用名为 writer 的变量。use 函数将接受带有单个参数的 lambda 表达式，如下所示。

**代码:**

```
fun main (args: Array<String>) {
FileWriter ("kotlin.txt")
  .use { w -> w.write ("kotlin") }
}
```

**输出:**

![kotlin try-with-resources 5](../Images/f9d025f97c99bffef0bd7c476875c1d3.png)

<noscript><img class="alignnone wp-image-545678 size-full" src="../Images/f9d025f97c99bffef0bd7c476875c1d3.png" alt="kotlin try-with-resources 5" width="520" height="228" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-5.jpg 520w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-5-300x132.jpg 300w" sizes="(max-width: 520px) 100vw, 520px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-5.jpg"/></noscript>

我们也可以在块内部使用隐式变量。

下面的例子显示了如何在块内部使用隐式变量，如下所示。

**代码:**

```
fun main(args: Array<String>) {
FileWriter ("kotlin.txt")
  .use { it.write ("kotlin") }
}
```

**输出:**

![implicit variable inside into the block](../Images/61275f3b7150e0251303486fde998dd5.png)

<noscript><img class="alignnone wp-image-545679 size-full" src="../Images/61275f3b7150e0251303486fde998dd5.png" alt="implicit variable inside into the block" width="560" height="223" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-6.jpg 560w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-6-300x119.jpg 300w" sizes="(max-width: 560px) 100vw, 560px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-6.jpg"/></noscript>

### Kotlin 使用资源尝试的示例

以下是 kotlin 使用以下资源进行尝试的示例:

#### 示例#1

在下面的例子中，我们使用变量名作为 ex 来定义 kotlin 中资源的例子。

**代码:**

```
fun main(args: Array<String>) {
  val ex = File("example.txt")
  ex.bufferedReader().use {
  println(it.readText())
  }
}
```

**输出:**

![kotlin try-with-resources 7](../Images/23e272450c5244f395f91d526d288631.png)

<noscript><img class="alignnone wp-image-545743 size-full" src="../Images/23e272450c5244f395f91d526d288631.png" alt="kotlin try-with-resources 7" width="545" height="264" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-7.jpg 545w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-7-300x145.jpg 300w" sizes="(max-width: 545px) 100vw, 545px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-7.jpg"/></noscript>

#### 实施例 2

下面的例子显示了我们如何在 kotlin 中使用 try with resource 来使用 use 函数，如下所示。

**代码:**

```
fun main(args: Array<String>) {
  val example = FileWriter ("example.txt")
  example.use {
  example.write ("Kotlin example")
  }
}
```

**输出:**

![use function](../Images/ac8f54288bb7f9a431a7d1e9bb9674fd.png)

<noscript><img class="alignnone wp-image-545751 size-full" src="../Images/ac8f54288bb7f9a431a7d1e9bb9674fd.png" alt="use function" width="577" height="273" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-8.jpg 577w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-8-300x142.jpg 300w" sizes="(max-width: 577px) 100vw, 577px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-8.jpg"/></noscript>

### 常见问题解答

下面是提到的常见问题:

#### Q1。kotlin try 中的 use 函数是如何使用资源的？

下面的代码显示了我们如何使用 kotlin try 资源的 use 函数，如下所示。

**代码:**

```
val func_use = FileWriter ("use.txt")
func_use.use
{
  func_use.write ("Use function")
}
```

**输出:**

![kotlin try-with-resources 9](../Images/b3fbeb623c496f1d1cfd96ce2eb57a88.png)

<noscript><img class="alignnone wp-image-545761 size-full" src="../Images/b3fbeb623c496f1d1cfd96ce2eb57a88.png" alt="kotlin try-with-resources 9" width="691" height="241" srcset="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-9.jpg 691w, https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-9-300x105.jpg 300w" sizes="(max-width: 691px) 100vw, 691px" data-original-src="https://cdn.educba.com/academy/wp-content/uploads/2022/07/kotlin-try-with-resources-9.jpg"/></noscript>

#### Q2。kotlin try with resource 中 use 函数的语法是什么？

**语法:**

```
Inline fun < >, use (block) R (source)
```

#### Q3。kotlin 使用资源尝试中的可关闭和自动关闭方法是什么？

**答案:**

我们正在用 kotlin 中的 closeable 方法定义 use 函数定义。java 中不存在自动关闭的方法。

### 结论

资源只不过是一个对象，在我们的程序中它被使用一次就被关闭了，例如套接字连接资源。Kotlin try with resources 用于管理 Kotlin 中的内存；科特林正在以一种非常有效的方式管理内存。

### 推荐文章

这是一个 Kotlin try-with-resources 指南。在这里，我们讨论介绍，kotlin 尝试与资源管理，例子和常见问题。您也可以看看以下文章，了解更多信息–

1.  [Kotlin Web 框架](https://www.educba.com/kotlin-web-framework/)
2.  [科特林内部](https://www.educba.com/kotlin-internal/)
3.  [科特林印花](https://www.educba.com/kotlin-print/)
4.  科特林队列

<footer class="entry-footer">

<aside class="sidebar sidebar-primary widget-area" role="complementary" aria-label="Primary Sidebar">Primary Sidebar</aside>

</footer>